File: python/graph.py

class Graph:
    def __init__(self):
        self.connections = {}

    @property
    def vertices(self):
        return self.connections.keys()

    @property
    def islands(self):
        return [i for i in self.connections.keys() if self.neighbors(i) == []]

    def add_edge(self, start, end, weight=1, bidirectional=True):
        if end in self.neighbors(start):
            self.purge(start, end)
        self.connections[start].append((end, weight))
        if bidirectional:
            self.connections[end].append((start, weight))

    def remove_edge(self, start, end, bidirectional=True):
        self.purge(start, end)
        if bidirectional:
            self.purge(end, start)

    def add_vertex(self):
        new = len(self.vertices)
        self.connections[new] = []

    def remove_vertex(self, index):
        self.connections.pop(index)
        if len(self.vertices) > 0:
            for i in self.vertices:
                self.purge(i, index)

    def connection(self, start, end):
        if start in self.vertices:
            return end in self.connections[start]
        else:
            return False

    def neighbors(self, start, with_weights=False):
        if with_weights:
            return self.connections[start]
        else:
            return [i[0] for i in self.connections[start]]

    def purge(self, start, target):
        self.connections[start] = [(j[0], j[1])
                                   for j in self.connections if j[0] != target]

    def weight(self, start, end):
        return [i[1] for i in self.connections[start] if i[0] == end][0]

    def bfs(self, vertex):
        explored = [vertex]
        paths = {vertex: 0}
        n = 0
        count = len(self.vertices) - len(self.islands)

        if vertex in self.islands:
            return False

        while len(explored) < count:
            start = explored[n]
            for end in self.neighbors(start):
                if end not in explored:
                    paths[end] = paths[start] + self.weight(start, end)
                    explored.append(end)
            n += 1
        return paths

    def dfs(self, vertex):
        explored = [vertex]
        paths = {vertex: 0}
        weight = 0

        def explore(self, vertex, length=0):
            neighbors = self.neighbors(vertex, True)
            for i in neighbors:
                if i[0] not in explored:
                    save = i[0]
                    length += self.weight(vertex, save)
                    explored.append(save)
                    paths[save] = length
                    explore(self, save, length)

        explore(self, vertex)
        return paths

    def dij(self, vertex):
        explored = [vertex]
        paths = {vertex: 0}
        count = len(self.vertices) - len(self.islands)

        lst = []
        while len(explored) < count:
            # neighbor scores
            for i in self.neighbors(vertex):
                lst.append((paths[vertex] + self.weight(vertex, i), i))
            lst = sorted(lst)
            # purge step
            lst2 = []
            record = set()
            for i, j in enumerate(lst):
                if j[1] in lst or j[1] in explored:
                    record.add(i)
            lst = [i[1] for i in enumerate(lst) if i[0] not in record]
            paths[lst[0][1]] = lst[0][0]
            vertex = lst[0][1]
            del lst[0]
            explored.append(vertex)
        return paths

File: java/Graph.java
import java.util.ArrayList;
import java.util.Hashtable;

public class Graph {
    private Hashtable<Integer, ArrayList<ArrayList<Integer>>> connections;

    public Graph() {
        connections = new Hashtable<Integer, ArrayList<ArrayList<Integer>>>();
    }

    public ArrayList<Integer> vertices() {
        return Collections.list(connections.keys());
    }

    public ArrayList<Integer> islands() {
        ArrayList<Integer> lst = new ArrayList<Integer>();
        for (int i : vertices()) {
            if (neighbors(i).size() == 0) {
                lst.add(i);
            }
        }
        return lst;
    }

    public void add_edge(int start, int end) {
        int weight = 0;
        boolean bidirectional = false;
        add_edge(start, end, weight, bidirectional);
    }

    public void add_edge(int start, int end, int weight) {
        boolean bidirectional = false;
        add_edge(start, end, weight, bidirectional);
    }

    public void add_edge(int start, int end, boolean bidirectional) {
        int weight = 0;
        add_edge(start, end, weight, bidirectional);
    }

    public void add_edge(int start, int end, int weight, boolean bidirectional) {
        if (connections.get(start).contains(end)) {
            purge(start, end);
        }
        connections.get(start).add(new ArrayList<Integer>(Arrays.asList(end, weight)));
        if (bidirectional) {
            connections.get(end).add(new ArrayList<Integer>(Arrays.asList(start, weight)));
        }
    }

    public void remove_edge(int start, int end) {
        boolean bidirectional = false;
        remove_edge(start, end, bidirectional);
    }

    public void remove_edge(int start, int end, boolean bidirectional) {
        purge(start, end);
        if (bidirectional) {
            purge(end, start);
        }
    }

    public void add_vertex() {
        connections.put(vertices().size(), new ArrayList<ArrayList<Integer>>());
    }

    public void remove_vertex(int index) {
        connections.remove(index);
        if (vertices().size() > 0) {
            for (int i : vertices()) {
                purge(i, index);
            }
        }

    }

    public ArrayList<Integer> neighbors(int start) {
        boolean weighted = false;
        return neighbors(start, weighted);
    }

    public ArrayList<Integer> neighbors(int start, boolean weighted) {
        if (weighted) {
            return connections.get(start);
        } else {
            ArrayList<Integer> a = new ArrayList<Integer>();
            for (ArrayList<Integer> i : connections.get(start)) {
                a.add(i.get(0));
            }
            return a;
        }
    }

    public boolean connection(int start, int end) {
        if (vertices().contains(start)) {
            return connections.get(start).contains(end);
        } else {
            return false;
        }
    }

    public void purge(int start, int target) {
        ArrayList<ArrayList<Integer>> lst = new ArrayList<ArrayList<Integer>>();
        for (ArrayList<Integer> j : connections.get(start)) {
            if (j.get(0) != target) {
                lst.add(new ArrayList<Integer>(Arrays.asList(j.get(0), j.get(1))));
            }
        }
        connections = lst;
    }

    public int weight(int start, int end) {
        for (ArrayList<Integer> j : connections.get(start)) {
            if (j.get(0) == end) {
                return j.get(1);
            }
        }
        return 0;
    }
}